### PR TITLE
refactor: Move options into `config/`, split files and move impls to cc

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

**Review by olgavrou — APPROVED** [review]

File: test/train-sets/ref/cb_bad_one_of_string.stdout [hunk 1]
```diff
@@ -1 +1 @@
-[critical] vw (options.h:345): Error: 'bad_type' is not a valid choice for option --cb_type. Please select from {dm, dr, ips, mtr, sm}
```
Comment (bassmang) [inline test/train-sets/ref/cb_bad_one_of_string.stdout:1]: Hopefully this change won't happen so much anymore :)

File: vowpalwabbit/config/options.cc [hunk 2]
```diff
@@ -77,7 +78,10 @@ bool options_i::add_parse_and_check_necessary(const option_group_definition& gro
     m_options[option->m_name] = option;
   }
 
-  return group.check_necessary_enabled(*this) && group.check_one_of();
+  auto necessary_enabled = group.check_necessary_enabled(*this);
+  if (necessary_enabled) { group.check_one_of(); }
```
Comment (bassmang) [inline vowpalwabbit/config/options.cc:82]: Are there not groups without any `necessary` options that we still want to check one_of, Eg most of the groups in parse_args

File: test/train-sets/ref/cb_bad_one_of_string.stdout (hunk 1 above)
Comment (jackgerrits) [reply test/train-sets/ref/cb_bad_one_of_string.stdout:1]: Haha yes I hope so!

File: vowpalwabbit/config/options.cc (hunk 2 above)
Comment (jackgerrits) [reply vowpalwabbit/config/options.cc:82]: Yeah they are checked in the other `add_and_parse` function. In the case of necessary we only check the `one_of` if the group was enabled. This is due to `cb_type` having different sets of allowed values.

**Review by bassmang — APPROVED** [review]

**Pull request merged**